<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>

<body>
    <script>
        const wait = (ms, ok) => {
            return new Promise((resolve, reject) => {
                setTimeout(() => {
                    if (ok) resolve(`過了${ms}毫秒`)
                    else reject(new Error(`錯誤，過了${ms}毫秒`))
                }, ms)
            })
        }

        // 如果沒有await會變成一坨
        // wait(2000, true)
        //     .then(result => {
        //         console.log('result')
        //     })
        //     .catch(error => {
        //         console.log('error')
        //     })

        // async 條件
        // await 等待promise 做完後執行後面的code
        // 條件一 await 後面只能放promise
        // 條件二 await 只能放在async的function
        const main = async () => {
            await wait(1000, true)
            console.log('a')
            await wait(2000, true)
            console.log('b')
        }
        // async function main() {}
        main()

        // promise + async 一啟用
        const mmain = async () => {
            try {
                // 變數 = await Promise
                // 變數會是resolve() 回傳的資料
                // 如果是Promise reject 會直接跳到 catch
                const result1 = await wait(1000, true)
                console.log(result1)
                const result2 = await wait(2000, true)
                console.log(result2)
                const result3 = await wait(2000, true)
                console.log(result3)
            } catch (error) {
                console.log('catch')
                console.log(error)
            }
        }

        mmain()
    </script>
</body>

</html>